mlflow/types/schema.py

```diff
@@ -122,34 +122,267 @@ def from_numpy_type(cls, np_type):
+                "Missing keys in Object JSON. Expected to find keys `properties` and `type`"
+            )
+        if kwargs["type"] != "object":
+            raise MlflowException("Type mismatch, Object expects `object` as the type")
+        return cls([Property.from_json_dict(**prop) for prop in kwargs["properties"]])
+
+        # This corresponds to line 276's comment
+        # for name, prop in kwargs["properties"].items():
+        #     # this should be in-place update
+        #     prop["name"] = name
+        # return cls([Property.from_json_dict(**prop) for prop in kwargs["properties"].values()])
+
+
+class Array:
+    """
+    Specification used to represent a json-convertible array.
+    """
+
+    def __init__(
+        self,
+        dtype: Union[DataType, ObjectType, str],
+    ) -> None:
+        try:
+            self._dtype = DataType[dtype] if isinstance(dtype, str) else dtype
+        except KeyError:
+            raise MlflowException(
+                f"Unsupported type '{dtype}', expected instance of DataType, Array, Object or "
+                f"one of {[t.name for t in DataType]}"
+            )
+        if not isinstance(self.dtype, (DataType, Object)):
+            raise TypeError(
+                "Expected mlflow.types.schema.Datatype, mlflow.types.schema.Object"
+                f"or str for the 'dtype' argument, but got {self.dtype.__class__}"
+            )
+
+    @property
+    def dtype(self) -> Union[DataType, ObjectType]:
+        """The array data type."""
+        return self._dtype
+
+    def __eq__(self, other) -> bool:
+        if isinstance(other, Array):
+            return self.dtype == other.dtype
+        return False
+
+    def to_dict(self):
+        items = self.dtype.name if isinstance(self.dtype, DataType) else self.dtype.to_dict()
+        return {"type": "array", "items": items}
+
+    @classmethod
+    def from_json_dict(cls, **kwargs):
+        """
+        Deserialize from a json loaded dictionary.
+        The dictionary is expected to contain `type` and
+        `items` keys.
+        """
+        if not {"items", "type"} <= set(kwargs.keys()):
+            raise MlflowException(
+                "Missing keys in Array JSON. Expected to find keys `items` and `type`"
+            )
+        if kwargs["type"] != "array":
+            raise MlflowException("Type mismatch, Array expects `array` as the type")
+        if isinstance(kwargs["items"], str):
+            return cls(dtype=kwargs["items"])
+        return cls(Object.from_json_dict(**kwargs["items"]))
+
+
 class ColSpec:
     """
     Specification of name and type of a single column in a dataset.
     """
 
     def __init__(
         self,
-        type: Union[DataType, str],  # pylint: disable=redefined-builtin
+        type: Union[DataType, Array, Object, str],  # pylint: disable=redefined-builtin
         name: Optional[str] = None,
-        optional: bool = False,
+        required: bool = True,
```

Comment:
Is this backwards compatible for validation of signatures of models logged pre/post this change?

Reply:
It's not. That's my question here, should we make a breaking change or keeps optional for some time and raise a deprecation warning?

Reply:
Sorry - I didn't see that message up above :) 
I am in favor of raising a deprecation warning, adding the required field. We might want to keep this around for quite some time in deprecated status due to serving environments being very outdated with training environments for our users.